### PR TITLE
Migrate tube lookup to V2 api

### LINKED ISSUE #634: BUG - Timeouts in GBS pool tube creation
**Describe the bug**
GBS MiSeq Pool tube creation was returning a 404 after 60s, due to the tube search taking too long to return. I've temporarily increased the timeouts in nginx to relieve the problem. But should do the following:

1) Add a 503 error template to avoid the confusion of a 404
2) Update to use the V2 tube endpoint to improve performance
3) Consider further improvements (such as asynchonous calls)

**RT Ticket Number**
#712129

**To Reproduce**
Steps to reproduce the behaviour:
1. Go to page '...'
2. Click on button '...'
3. See error

**Expected behaviour**
What was expected to happen.

**Screenshots**

**Desktop (please complete the following information):**
* Windows or Mac?
* Browser Chrome, Firefox, Safari or other?
* Browser version (use 'About' to look up)?

**Additional context**

### PR DESCRIPTION
**Pull request by JamesGlover**
The V1 API for tubes is incredibly slow, as it always renders the
full content of the tube. By contrast, V1 allows an utterly minimal
representation.

*Depends on* https://github.com/sanger/sequencescape/pull/3075

Closes #634

Changes proposed in this pull request:

*
*
* ...

### CODE REVIEW

File: app/models/labware_creators/pooled_tubes_from_whole_tubes.rb [hunk 1]
```diff
@@ -11,7 +11,6 @@ class SubmissionFailure < StandardError; end
 
     self.page = 'pooled_tubes_from_whole_tubes'
     self.attributes += [{ barcodes: [] }]
-    self.default_transfer_template_name = 'Whole plate to tube'
```
Comment (JamesGlover) [inline app/models/labware_creators/pooled_tubes_from_whole_tubes.rb:14]: Not relevant for this creator

File: app/models/labware_creators/pooled_tubes_from_whole_tubes.rb [hunk 2]
```diff
@@ -39,33 +38,28 @@ def barcodes=(input)
...
       errors.add(:barcodes, "could not be found: #{missing_barcodes}") unless missing_barcodes.empty?
     end
 
     private
 
     def transfer_request_attributes
-      parents.each_with_object([]) do |parent, transfer_requests|
-        transfer_requests << { source_asset: parent.uuid, target_asset: @child.uuid }
+      parents.map do |parent|
+        { source_asset: parent.uuid, target_asset: @child.uuid }
```
Comment (JamesGlover) [inline app/models/labware_creators/pooled_tubes_from_whole_tubes.rb:61]: No idea why this was each with object.

File: app/models/labware_creators/pooled_tubes_from_whole_tubes.rb [hunk 3]
```diff
@@ -39,33 +38,28 @@ def barcodes=(input)
...
-                                                                                    barcode: barcodes)
+      @parents ||= Sequencescape::Api::V2::Tube.find_all({ barcode: barcodes }, includes: [])
     end
 
     def parents_suitable
-      missing_barcodes = barcodes - parents.map { |p| p.barcode.ean13 }
+      # Plate#barcode =~ ensures different 'flavours' of the same barcode still match.
+      # Ie. EAN13 encoded versions will match the Code39 encoded versions.
+      missing_barcodes = barcodes.reject { |scanned_bc| parents.any? { |p| p.barcode =~ scanned_bc } }
       errors.add(:barcodes, "could not be found: #{missing_barcodes}") unless missing_barcodes.empty?
```
Comment (JamesGlover) [inline app/models/labware_creators/pooled_tubes_from_whole_tubes.rb:54]: In practice tubes are currently ean13, but this kind of code bit us when we updated plates to code39.

File: script/limber [hunk 4]
```diff
@@ -53,7 +54,7 @@ begin
   $stdout.puts 'Building configuration'
   $stdout.puts `bundle exec rake config:generate`
   $stdout.puts 'Running application'
-  exec('bundle exec rails s')
+  exec("bundle exec rails #{command}")
 rescue Errno::ENOENT
```
Comment (JamesGlover) [inline script/limber:58]: `./script/limber` lets you preconfigure limber to talk to different environments with `./script/limber ENVIRONMENT` this lets you also start up a rails console, via `./script/limber ENVIRONMENT c`

File: app/models/labware_creators/pooled_tubes_from_whole_tubes.rb (hunk 1 above)
Comment (KatyTaylor) [reply app/models/labware_creators/pooled_tubes_from_whole_tubes.rb:14]: Just noticed the comment at the top of this file says `Pools an entire plate into a single tube. Useful for MiSeqQC`
Presume should say `Pools an entire tube into a single tube. Useful for MiSeqQC`

File: app/models/labware_creators/pooled_tubes_from_whole_tubes.rb [hunk 5]
```diff
@@ -39,33 +38,28 @@ def barcodes=(input)
 
     # TODO: This should probably be asynchronous
     def available_tubes
```
Comment (KatyTaylor) [inline app/models/labware_creators/pooled_tubes_from_whole_tubes.rb:40]: Does this method find tubes that are appropriate to be the source tubes? (correct purpose)

File: app/models/labware_creators/pooled_tubes_from_whole_tubes.rb [hunk 6]
```diff
@@ -39,33 +38,28 @@ def barcodes=(input)
...
-      @search_options = OngoingTube.new(purposes: [parent.purpose.uuid], include_used: false, states: ['passed'])
-      @search_results = tube_search.all(
-        Limber::Tube,
-        @search_options.search_parameters
-      )
+      @search_options = OngoingTube.new(purpose_names: [parent.purpose.name], include_used: false)
```
Comment (KatyTaylor) [inline app/models/labware_creators/pooled_tubes_from_whole_tubes.rb:41]: Why remove the `states: ['passed']` part?

Comment (JamesGlover) [reply app/models/labware_creators/pooled_tubes_from_whole_tubes.rb:41]: The existing search functionality was actually ignoring the filter anyway, but it occurred to me that its probably quite useful to see tubes that are almost ready as well. A pending or started tube in this list is only one step away from being available.

File: app/models/labware_creators/pooled_tubes_from_whole_tubes.rb (hunk 5 above)
Comment (JamesGlover) [reply app/models/labware_creators/pooled_tubes_from_whole_tubes.rb:40]: Yep, it tries to exclude those that have been used, but there's another bug elsewhere in Limber that means it isn't always 100% successful in this. (I've added a story here as it seemed to be missing https://github.com/sanger/limber/issues/645)

I also noticed that there seem to be a few tubes 'abandoned' at this stage. Some of them were completed on another fork, others looks just genuinely abandoned. The former we could probably add a scope to filter out, but the latter will always hand about.

There's a limit to confine the returned tubes to the 30 most recent though, so this list shouldn't get too unmanageable.

File: app/models/labware_creators/pooled_tubes_from_whole_tubes.rb [hunk 7]
```diff
@@ -39,33 +38,28 @@ def barcodes=(input)
...
-      @parents ||= api.search.find(Settings.searches['Find assets by barcode']).all(Limber::BarcodedAsset,
-                                                                                    barcode: barcodes)
+      @parents ||= Sequencescape::Api::V2::Tube.find_all({ barcode: barcodes }, includes: [])
     end
 
     def parents_suitable
-      missing_barcodes = barcodes - parents.map { |p| p.barcode.ean13 }
+      # Plate#barcode =~ ensures different 'flavours' of the same barcode still match.
+      # Ie. EAN13 encoded versions will match the Code39 encoded versions.
+      missing_barcodes = barcodes.reject { |scanned_bc| parents.any? { |p| p.barcode =~ scanned_bc } }
```
Comment (KatyTaylor) [inline app/models/labware_creators/pooled_tubes_from_whole_tubes.rb:53]: Bit confused by this line. Looks to me like `=~` checks if the string one side is contained within the other, and returns either the offset (integer) or nil. But I would have thought `.any?` wants a boolean returned..?

File: app/views/sequencescape/api/v2/tubes/_tube.html.erb [hunk 8]
```diff
@@ -0,0 +1,10 @@
+<li class="labware">
+  <%= link_to(tube, id: "tube_#{tube.barcode.ean13}", class: "list-group-item-action") do %>
+    <h3 class="barcode list-group-item-heading"><%= useful_barcode(tube.barcode) %> <%= state_badge(tube.state) %></h3>
+    <p><%= tube.name %>
+    <dl>
+      <dt>Created</dt><dd><%= tube.created_at.to_formatted_s(:date_created) %></dd>
+      <dt>Type</dt><dd><%= tube.purpose.name %></dd>
+    </dl>
+  <% end -%>
+</li>
```
Comment (KatyTaylor) [inline app/views/sequencescape/api/v2/tubes/_tube.html.erb:10]: where will this be displayed?

File: app/models/labware_creators/pooled_tubes_from_whole_tubes.rb (hunk 1 above)
Comment (JamesGlover) [reply app/models/labware_creators/pooled_tubes_from_whole_tubes.rb:14]: Actually that line should be completely re-written, its left over from another creator that I used to template this one.

File: app/models/labware_creators/pooled_tubes_from_whole_tubes.rb (hunk 7 above)
Comment (JamesGlover) [reply app/models/labware_creators/pooled_tubes_from_whole_tubes.rb:53]: Barcode in this case isn't a string, but a LabwareBarcode object. https://github.com/sanger/limber/blob/87cb3fbb8ceb0ba4f102b8dc3831ee23814f36f8/lib/labware_barcode.rb

Comment (KatyTaylor) [reply app/models/labware_creators/pooled_tubes_from_whole_tubes.rb:53]: ah it's got a special thingamybob

**Review by KatyTaylor — APPROVED** [review]

File: app/views/sequencescape/api/v2/tubes/_tube.html.erb (hunk 8 above)
Comment (JamesGlover) [reply app/views/sequencescape/api/v2/tubes/_tube.html.erb:10]: `app/views/tube_creation/pooled_tubes_from_whole_tubes.html.erb` renders the partial:
`app/views/search/_inbox.html.erb`
Which contains
`render sorted_labware`
When render receives an array, rails treats it as a collection, and selects the template to used based on the class name. (More accurately various helper methods provided by ActiveRecord/ActiveModel)
In this case we were using the default template for the V1 API stuff, and this just sets up a similar template for the V2 stuff, allowing rails to switch over automatically.

Comment (JamesGlover) [reply app/views/sequencescape/api/v2/tubes/_tube.html.erb:10]: I've not removed the V1 templates yet, as we'll need to migrate over the inboxes first. (Which we should definitely do)

File: app/models/labware_creators/pooled_tubes_from_whole_tubes.rb (hunk 1 above)
Comment (JamesGlover) [reply app/models/labware_creators/pooled_tubes_from_whole_tubes.rb:16]: Updated the comment to remove references to plates (And MiSeq QC)

File: app/models/labware_creators/pooled_tubes_from_whole_tubes.rb (hunk 7 above)
Comment (JamesGlover) [reply app/models/labware_creators/pooled_tubes_from_whole_tubes.rb:55]: Wish I hadn't used the standard match operator, as it caused confusion to Andrew as well. I'll probably rename that at somepoint, or at least add an alias.

File: app/views/sequencescape/api/v2/tubes/_tube.html.erb (hunk 8 above)
Comment (KatyTaylor) [reply app/views/sequencescape/api/v2/tubes/_tube.html.erb:10]: Cool OK thanks
Do you think it's worth a comment in this file with an abridged version of that? Just thinking if someone comes across it while on a deleting spree it's not going to be searchable where it's used

**Pull request merged**